src/main/java/com/openshift/internal/restclient/ApiTypeMapper.java

```diff
@@ -0,0 +1,389 @@
+	}
+	
+	private synchronized void init() {
+		if(resourceEndpoints != null) {
+			return;
+		}
+		resourceEndpoints = new ArrayList<>();
+		Collection<ApiGroup> groups = getLegacyGroups();
+		groups.addAll(getApiGroups());
+		groups.forEach(g->{
+			Collection<String> versions = g.getVersions();
+			versions.forEach(v->{
+				Collection<ModelNode> resources = getResources(g, v);
+				addEndpoints(g.getPrefix(), g.getName(), v, resources);
+			});
+		});
+	}
+	
+	private void addEndpoints(final String prefix, final String apiGroupName, final String version, final Collection<ModelNode> nodes) {
+		for (ModelNode node : nodes) {
+			String name = node.get(NAME).asString();
+			String capability = null;
+			if(name.contains(FWD_SLASH)) {
+				int first = name.indexOf(FWD_SLASH);
+				capability = name.substring(first+1);
+				name = name.substring(0, first);
+			}
+			boolean namespaced = node.get("namespaced").asBoolean();
+			VersionedApiResource resource = new VersionedApiResource(prefix, apiGroupName, version, name, node.get(KIND).asString(), namespaced);
+			if(!resourceEndpoints.contains(resource)) {
+				resourceEndpoints.add(resource);
+			}
+			if(capability != null) {
+				int index = resourceEndpoints.indexOf(resource);
+				resourceEndpoints.get(index).addCapability(capability);
+			}
+		}
+	}
+
+	private Collection<ApiGroup> getApiGroups(){
+		String json = readEndpoint(API_GROUPS_API);
+		return ModelNode.fromJSONString(json)
+				.get("groups")
+				.asList()
+				.stream()
+				.map(n->new ApiGroup(API_GROUPS_API, n))
+				.collect(Collectors.toList());
+	}
+	
+	private Collection<ModelNode> getResources(IApiGroup group, String version){
+		String json = readEndpoint(group.pathFor(version));
+		if(StringUtils.isBlank(json)) {
+			return new ArrayList<>();
+		}
+		ModelNode node = ModelNode.fromJSONString(json);
+		return node.get("resources").asList();
+	}
+	
+	private Collection<ApiGroup> getLegacyGroups(){
+		Collection<ApiGroup> groups = new ArrayList<>();
+		for(String e: Arrays.asList(KUBE_API,OS_API)) {
+			String json = readEndpoint(e);
+			ModelNode n = ModelNode.fromJSONString(json);
+			groups.add(new LegacyApiGroup(e,n));
+		}
+		return groups;
+	}
+	
+	private String readEndpoint(final String endpoint) {
+		try {
+			final URL url = new URL(new URL(this.baseUrl), endpoint);
+			LOGGER.debug(url.toString());
+			String response = client.get(url, IHttpConstants.DEFAULT_READ_TIMEOUT);
+			LOGGER.debug(response);
+			return response;
+		} catch (MalformedURLException | SocketTimeoutException e) {
+			LOGGER.error("Exception", e);
+			throw new OpenShiftException(e,"");
+		//HACK - This gets us around a server issue
+		} catch (HttpClientException e) {
+			if(e instanceof NotFoundException) {
+				throw new com.openshift.restclient.NotFoundException(e);
+			}
+			LOGGER.error("Unauthorized exception. Can system:anonymous get the API endpoint", e);
```

Comment:
here i will defer not replacing unless we want to wrap up the exception in openshiftexception with the message that we are logging here.